// Copyright (c) 2016-present, Facebook, Inc.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

syntax = "proto3";

package magma.migration.cellular;
option go_package = "types";

message LegacySubscriberID {
    string id = 1;

    enum IDType {
        IMSI = 0;
    }
    IDType type = 2;
}

message GSMSubscription {
    enum GSMSubscriptionState {
        INACTIVE = 0;
        ACTIVE = 1;
    }
    GSMSubscriptionState state = 1;

    enum GSMAuthAlgo {
        PRECOMPUTED_AUTH_TUPLES = 0;
    }
    GSMAuthAlgo auth_algo = 2;
    bytes auth_key = 3;
    repeated bytes auth_tuples = 4;
}

message LegacyLTESubscription {
    enum LTESubscriptionState {
        INACTIVE = 0;
        ACTIVE = 1;
    }
    LTESubscriptionState state = 1;

    enum LTEAuthAlgo {
        MILENAGE = 0;
    }
    LTEAuthAlgo auth_algo = 2;
    bytes auth_key = 3;
    bytes auth_opc = 4;
}

message LegacySubscriberState {
    uint64 lte_auth_next_seq = 1;
    string tgpp_aaa_server_name = 2;
    bool tgpp_aaa_server_registered = 3;
}

enum AccessNetworkIdentifier {
    HRPD = 0;
    WIMAX = 1;
    WLAN = 2;
    ETHERNET = 3;
}

message APNConfiguration {
    uint32 context_id = 1;
    string service_selection = 2;
    QoSProfile qos_profile = 3;
    AggregatedMaximumBitrate ambr = 4;

    enum PDNType {
        IPV4 = 0;
        IPV6 = 1;
        IPV4V6 = 2;
        IPV4_OR_IPV6 = 3;
    }
    PDNType pdn = 5;

    message QoSProfile {
        int32 class_id = 1;
        uint32 priority_level = 2;
        bool preemption_capability = 3;
        bool preemption_vulnerability = 4;
    }
}

message AggregatedMaximumBitrate {
    uint32 max_bandwidth_ul = 1;
    uint32 max_bandwidth_dl = 2;
}

message Non3GPPUserProfile {
    string msisdn = 1;

    enum Non3GPPIPAccess {
        NON_3GPP_SUBSCRIPTION_ALLOWED = 0;
        NON_3GPP_SUBSCRIPTION_BARRED = 1;
    }
    Non3GPPIPAccess non_3gpp_ip_access = 2;

    enum Non3GPPIPAccessAPN {
        NON_3GPP_APNS_ENABLE = 0;
        NON_3GPP_APNS_DISABLE = 1;
    }
    Non3GPPIPAccessAPN non_3gpp_ip_access_apn = 3;
    AggregatedMaximumBitrate ambr = 4;
    APNConfiguration apn_config = 5;
    AccessNetworkIdentifier access_net_id = 6;
}

// --------------------------------------------------------------------------
// User subscription record. This record is serialized and stored in the DB.
// --------------------------------------------------------------------------
message SubscriberData {
    LegacySubscriberID sid = 1;
    GSMSubscription gsm = 2;
    LegacyLTESubscription lte = 3;
    NetworkID network_id = 4;
    LegacySubscriberState state = 5;
    string sub_profile = 6;
    Non3GPPUserProfile non_3gpp = 7;
}

message NetworkID {
    string id = 1;
}
